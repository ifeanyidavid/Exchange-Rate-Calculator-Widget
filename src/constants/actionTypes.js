export const FETCH_RATES_START = "FETCH_RATES_START";
export const FETCH_RATES_ERROR = "FETCH_RATES_ERROR";
export const FETCH_RATES_DONE = "FETCH_RATES_DONE";
export const UPDATE_SOURCE_CURRENCY = "UPDATE_SOURCE_CURRENCY";
export const UPDATE_DESTINATION_CURRENCY = "UPDATE_DESTINATION_CURRENCY";
export const EXCHANGE_CURRENCY = "EXCHANGE_CURRENCY";
