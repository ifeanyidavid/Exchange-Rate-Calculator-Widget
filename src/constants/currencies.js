export const currencies = {
  EUR: { id: 0, value: "EUR" },
  GBP: { id: 1, value: "GBP" },
  USD: { id: 2, value: "USD" }
};
