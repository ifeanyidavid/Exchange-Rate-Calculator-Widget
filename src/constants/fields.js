export const fieldNames = {
  sourceCurrency: "sourceCurrency",
  destinationCurrency: "destinationCurrency",
  sourceAmount: "sourceAmount",
  destinationAmount: "destinationAmount",
  balance: "balance",
  symbol: "symbol",
  description: "description",
  currency: "currency",
  id: "id",
  error: "error",
  loading: "loading",
  rates: "rates",
};
