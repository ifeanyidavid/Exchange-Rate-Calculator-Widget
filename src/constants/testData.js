import { currencies } from "./currencies";
import { fieldNames } from "./fields";

export const testData = {
  rates: [
    {
      rates: { USD: 1.1075, GBP: 0.84195 },
      base: "EUR",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 1.1877189857, USD: 1.3153987766 },
      base: "GBP",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 0.9029345372, GBP: 0.7602257336 },
      base: "USD",
      date: "2019-12-09"
    }
  ],
  pocketsReducer: {
    transactionHistory: [
      {
        id: 0,
        description: `Exchanged to EUR`,
        time: new Date().toLocaleTimeString("en-GB"),
        sourceAmount: 23,
        destinationAmount: 42,
        sourceCurrency: "USD",
        destinationCurrency: "EUR"
      }
    ],
    pockets: {
      [currencies.EUR.value]: {
        [fieldNames.id]: currencies.EUR.id,
        [fieldNames.currency]: currencies.EUR.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "€",
        [fieldNames.description]: "Euro"
      },
      [currencies.GBP.value]: {
        [fieldNames.id]: currencies.GBP.id,
        [fieldNames.currency]: currencies.GBP.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "£",
        [fieldNames.description]: "British Pound"
      },
      [currencies.USD.value]: {
        [fieldNames.id]: currencies.USD.id,
        [fieldNames.currency]: currencies.USD.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "$",
        [fieldNames.description]: "American Dollar"
      }
    }
  },
  sourceAmount: 23,
  destinationAmount: 54,
  sourceCurrency: "USD",
  destinationCurrency: "EUR",
  error: new Error("Network disconnected"),
  slideIndex: 1,
  pocketValues: [
    {
      [fieldNames.id]: currencies.EUR.id,
      [fieldNames.currency]: currencies.EUR.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "€",
      [fieldNames.description]: "Euro"
    },
    {
      [fieldNames.id]: currencies.GBP.id,
      [fieldNames.currency]: currencies.GBP.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "£",
      [fieldNames.description]: "British Pound"
    },
    {
      [fieldNames.id]: currencies.USD.id,
      [fieldNames.currency]: currencies.USD.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "$",
      [fieldNames.description]: "American Dollar"
    }
  ]
};
