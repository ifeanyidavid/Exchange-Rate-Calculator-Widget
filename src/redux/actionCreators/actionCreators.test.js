import configureMockStore from "redux-mock-store";

import {
  saveRatesStart,
  saveRatesFail,
  saveRatesDone,
  exchangeCurrency,
  updateDestinationCurrency,
  updateSourceCurrency
} from "./index";
import { testData } from "../../constants/testData";

const mockStore = configureMockStore();
const store = mockStore({});

/* RATES ACTION CREATORS TEST CASES */

describe("Rates action creators", () => {
  it("should dispatch FETCH_RATES_START action", () => {
    store.dispatch(saveRatesStart());
    expect(store.getActions()).toMatchSnapshot();
  });

  it("should dispatch FETCH_RATES_ERROR and error to store in state", () => {
    store.dispatch(saveRatesFail(testData.error));
    expect(store.getActions()).toMatchSnapshot();
  });

  it("should dispatch FETCH_RATES_DONE and data to store in state", () => {
    store.dispatch(saveRatesDone(testData.rates));
    expect(store.getActions()).toMatchSnapshot();
  });

  it("should dispatch UPDATE_SOURCE_CURRENCY and slideIndex to store in state", () => {
    store.dispatch(updateSourceCurrency(testData.slideIndex));
    expect(store.getActions()).toMatchSnapshot();
  });

  it("should dispatch UPDATE_DESTINATION_CURRENCY and slideIndex to store in state", () => {
    store.dispatch(updateDestinationCurrency(testData.slideIndex));
    expect(store.getActions()).toMatchSnapshot();
  });
});

/* POCKETS ACTION CREATORS TEST CASES */
describe("Pockets action creators", () => {
  it("should EXCHANGE_CURRENCY and data to store in state", () => {
    const sourceAmount = testData.sourceAmount;
    const destinationAmount = testData.destinationAmount;
    const sourceCurrency = testData.sourceCurrency;
    const destinationCurrency = testData.destinationCurrency;

    store.dispatch(
      exchangeCurrency(
        sourceAmount,
        destinationAmount,
        sourceCurrency,
        destinationCurrency
      )
    );

    expect(store.getActions()).toMatchSnapshot();
  });
});
