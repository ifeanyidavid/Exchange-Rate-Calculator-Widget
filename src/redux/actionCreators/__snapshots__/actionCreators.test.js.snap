// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pockets action creators should EXCHANGE_CURRENCY and data to store in state 1`] = `
Array [
  Object {
    "type": "FETCH_RATES_START",
  },
  Object {
    "payload": [Error: Network disconnected],
    "type": "FETCH_RATES_ERROR",
  },
  Object {
    "payload": Array [
      Object {
        "base": "EUR",
        "date": "2019-12-09",
        "rates": Object {
          "GBP": 0.84195,
          "USD": 1.1075,
        },
      },
      Object {
        "base": "GBP",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 1.1877189857,
          "USD": 1.3153987766,
        },
      },
      Object {
        "base": "USD",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 0.9029345372,
          "GBP": 0.7602257336,
        },
      },
    ],
    "type": "FETCH_RATES_DONE",
  },
  Object {
    "payload": 1,
    "type": "UPDATE_SOURCE_CURRENCY",
  },
  Object {
    "payload": 1,
    "type": "UPDATE_DESTINATION_CURRENCY",
  },
  Object {
    "payload": Object {
      "destinationAmount": 54,
      "destinationCurrency": "EUR",
      "sourceAmount": 23,
      "sourceCurrency": "USD",
    },
    "type": "EXCHANGE_CURRENCY",
  },
]
`;

exports[`Rates action creators should dispatch FETCH_RATES_DONE and data to store in state 1`] = `
Array [
  Object {
    "type": "FETCH_RATES_START",
  },
  Object {
    "payload": [Error: Network disconnected],
    "type": "FETCH_RATES_ERROR",
  },
  Object {
    "payload": Array [
      Object {
        "base": "EUR",
        "date": "2019-12-09",
        "rates": Object {
          "GBP": 0.84195,
          "USD": 1.1075,
        },
      },
      Object {
        "base": "GBP",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 1.1877189857,
          "USD": 1.3153987766,
        },
      },
      Object {
        "base": "USD",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 0.9029345372,
          "GBP": 0.7602257336,
        },
      },
    ],
    "type": "FETCH_RATES_DONE",
  },
]
`;

exports[`Rates action creators should dispatch FETCH_RATES_ERROR and error to store in state 1`] = `
Array [
  Object {
    "type": "FETCH_RATES_START",
  },
  Object {
    "payload": [Error: Network disconnected],
    "type": "FETCH_RATES_ERROR",
  },
]
`;

exports[`Rates action creators should dispatch FETCH_RATES_START action 1`] = `
Array [
  Object {
    "type": "FETCH_RATES_START",
  },
]
`;

exports[`Rates action creators should dispatch UPDATE_DESTINATION_CURRENCY and slideIndex to store in state 1`] = `
Array [
  Object {
    "type": "FETCH_RATES_START",
  },
  Object {
    "payload": [Error: Network disconnected],
    "type": "FETCH_RATES_ERROR",
  },
  Object {
    "payload": Array [
      Object {
        "base": "EUR",
        "date": "2019-12-09",
        "rates": Object {
          "GBP": 0.84195,
          "USD": 1.1075,
        },
      },
      Object {
        "base": "GBP",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 1.1877189857,
          "USD": 1.3153987766,
        },
      },
      Object {
        "base": "USD",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 0.9029345372,
          "GBP": 0.7602257336,
        },
      },
    ],
    "type": "FETCH_RATES_DONE",
  },
  Object {
    "payload": 1,
    "type": "UPDATE_SOURCE_CURRENCY",
  },
  Object {
    "payload": 1,
    "type": "UPDATE_DESTINATION_CURRENCY",
  },
]
`;

exports[`Rates action creators should dispatch UPDATE_SOURCE_CURRENCY and slideIndex to store in state 1`] = `
Array [
  Object {
    "type": "FETCH_RATES_START",
  },
  Object {
    "payload": [Error: Network disconnected],
    "type": "FETCH_RATES_ERROR",
  },
  Object {
    "payload": Array [
      Object {
        "base": "EUR",
        "date": "2019-12-09",
        "rates": Object {
          "GBP": 0.84195,
          "USD": 1.1075,
        },
      },
      Object {
        "base": "GBP",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 1.1877189857,
          "USD": 1.3153987766,
        },
      },
      Object {
        "base": "USD",
        "date": "2019-12-09",
        "rates": Object {
          "EUR": 0.9029345372,
          "GBP": 0.7602257336,
        },
      },
    ],
    "type": "FETCH_RATES_DONE",
  },
  Object {
    "payload": 1,
    "type": "UPDATE_SOURCE_CURRENCY",
  },
]
`;
